File: generate-logs.py
#!/usr/bin/env python3
"""
Spring Boot JSON Lines Log Generator

This script generates sample Spring Boot logs in JSON Lines format
for testing the Loki + Grafana stack.
"""

import json
import random
import time
from datetime import datetime, timezone
import os
import sys

# Log levels with relative frequencies
LOG_LEVELS = [
    ("INFO", 60),
    ("DEBUG", 20),
    ("WARN", 15),
    ("ERROR", 4),
    ("TRACE", 1),
]

# Sample loggers
LOGGERS = [
    "com.example.demo.controller.UserController",
    "com.example.demo.service.UserService",
    "com.example.demo.repository.UserRepository",
    "com.example.demo.controller.OrderController",
    "com.example.demo.service.OrderService",
    "com.example.demo.service.PaymentService",
    "com.example.demo.security.AuthenticationFilter",
    "com.example.demo.config.DataSourceConfig",
    "org.springframework.web.servlet.DispatcherServlet",
    "org.springframework.boot.web.embedded.tomcat.TomcatWebServer",
]

# Sample messages by log level
MESSAGES = {
    "INFO": [
        "Application started successfully",
        "User logged in successfully",
        "Order created with ID: {}",
        "Payment processed successfully",
        "Database connection established",
        "Request completed in {}ms",
        "New user registered: {}",
        "Session created for user: {}",
        "Cache refreshed successfully",
        "Health check passed",
    ],
    "DEBUG": [
        "Entering method: {}",
        "Exiting method: {}",
        "Query executed: SELECT * FROM users WHERE id = {}",
        "Cache hit for key: {}",
        "Validating request parameters",
        "Processing request from IP: {}",
        "Applying security filter",
        "Deserializing JSON payload",
        "Loading configuration from: {}",
        "Initializing bean: {}",
    ],
    "WARN": [
        "Slow query detected: took {}ms",
        "Cache miss for key: {}",
        "Deprecated API used: {}",
        "Retry attempt {} for operation",
        "Queue size approaching limit: {}",
        "Connection pool utilization high: {}%",
        "Session timeout for user: {}",
        "Invalid input received: {}",
        "Rate limit approaching for IP: {}",
        "Configuration value missing, using default",
    ],
    "ERROR": [
        "Failed to process payment: {}",
        "Database connection error: {}",
        "Authentication failed for user: {}",
        "Unable to send email notification",
        "External API call failed: {}",
        "Invalid JSON payload received",
        "Resource not found: {}",
        "Permission denied for user: {}",
        "Transaction rollback: {}",
        "Unexpected exception: {}",
    ],
    "TRACE": [
        "Method trace: {} with parameters: {}",
        "SQL statement: {}",
        "HTTP request headers: {}",
        "Request body: {}",
        "Response body: {}",
    ],
}

# Sample thread names
THREADS = [
    "http-nio-8080-exec-1",
    "http-nio-8080-exec-2",
    "http-nio-8080-exec-3",
    "http-nio-8080-exec-4",
    "scheduling-1",
    "task-executor-1",
    "task-executor-2",
]

# Sample exception classes
EXCEPTION_CLASSES = [
    "java.sql.SQLException",
    "org.springframework.web.client.HttpClientErrorException",
    "java.io.IOException",
    "java.lang.NullPointerException",
    "javax.validation.ValidationException",
    "org.springframework.security.access.AccessDeniedException",
    "com.example.demo.exception.ResourceNotFoundException",
    "com.example.demo.exception.PaymentException",
]

# Sample exception messages
EXCEPTION_MESSAGES = [
    "Connection timeout after 30000ms",
    "Invalid credentials provided",
    "Resource with ID {} not found",
    "Payment gateway returned error code: {}",
    "Validation failed for field: {}",
    "Database constraint violation",
    "API rate limit exceeded",
    "Session expired",
]


def generate_trace_id():
    """Generate a random trace ID"""
    return ''.join(random.choices('0123456789abcdef', k=32))


def generate_span_id():
    """Generate a random span ID"""
    return ''.join(random.choices('0123456789abcdef', k=16))


def format_message(template, level):
    """Format a message template with random values"""
    if '{}' in template:
        if 'ID' in template or 'id' in template:
            return template.format(random.randint(1000, 9999))
        elif 'ms' in template:
            return template.format(random.randint(50, 2000))
        elif 'user' in template.lower():
            return template.format(f"user{random.randint(1, 100)}")
        elif '%' in template:
            return template.format(random.randint(70, 95))
        elif 'IP' in template:
            return template.format(f"192.168.1.{random.randint(1, 255)}")
        else:
            return template.format(random.choice(['alpha', 'beta', 'gamma', 'delta']))
    return template


def generate_log_entry():
    """Generate a single log entry in JSON Lines format"""
    # Select log level based on frequency
    level = random.choices(
        [l[0] for l in LOG_LEVELS],
        weights=[l[1] for l in LOG_LEVELS],
        k=1
    )[0]
    
    # Generate basic log entry
    log_entry = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "level": level,
        "thread": random.choice(THREADS),
        "logger": random.choice(LOGGERS),
        "message": format_message(random.choice(MESSAGES[level]), level),
        "application": "demo-app",
    }
    
    # Add trace and span IDs (50% of the time)
    if random.random() < 0.5:
        log_entry["traceId"] = generate_trace_id()
        log_entry["spanId"] = generate_span_id()
    
    # Add exception details for ERROR logs (70% of the time)
    if level == "ERROR" and random.random() < 0.7:
        exception_class = random.choice(EXCEPTION_CLASSES)
        exception_message = format_message(random.choice(EXCEPTION_MESSAGES), level)
        
        log_entry["exception"] = {
            "class": exception_class,
            "message": exception_message,
            "stackTrace": f"{exception_class}: {exception_message}\n\tat com.example.demo.Example.method(Example.java:{random.randint(10, 200)})\n\tat com.example.demo.Main.run(Main.java:{random.randint(10, 100)})"
        }
    
    return log_entry


def write_log(log_entry, log_file):
    """Write a log entry to file in JSON Lines format"""
    with open(log_file, 'a') as f:
        f.write(json.dumps(log_entry) + '\n')


def main():
    """Main function to generate logs"""
    # Create logs directory if it doesn't exist
    log_dir = "logs"
    if not os.path.exists(log_dir):
        os.makedirs(log_dir)
        print(f"Created directory: {log_dir}")
    
    log_file = os.path.join(log_dir, "application.log")
    
    # Determine mode: batch or continuous
    mode = sys.argv[1] if len(sys.argv) > 1 else "batch"
    
    if mode == "continuous":
        print(f"Generating logs continuously to {log_file}")
        print("Press Ctrl+C to stop")
        try:
            while True:
                log_entry = generate_log_entry()
                write_log(log_entry, log_file)
                
                # Print to console
                level_color = {
                    "INFO": "\033[32m",    # Green
                    "DEBUG": "\033[36m",   # Cyan
                    "WARN": "\033[33m",    # Yellow
                    "ERROR": "\033[31m",   # Red
                    "TRACE": "\033[37m",   # White
                }
                reset_color = "\033[0m"
                
                print(f"{level_color.get(log_entry['level'], '')}"
                      f"[{log_entry['level']:5s}] "
                      f"{log_entry['logger']:50s} - "
                      f"{log_entry['message']}"
                      f"{reset_color}")
                
                # Random delay between 0.1 and 2 seconds
                time.sleep(random.uniform(0.1, 2.0))
        except KeyboardInterrupt:
            print("\n\nStopped log generation")
    
    elif mode == "batch":
        # Generate a batch of logs
        count = int(sys.argv[2]) if len(sys.argv) > 2 else 100
        print(f"Generating {count} log entries to {log_file}")
        
        for i in range(count):
            log_entry = generate_log_entry()
            write_log(log_entry, log_file)
            
            if (i + 1) % 10 == 0:
                print(f"Generated {i + 1}/{count} logs")
        
        print(f"\nCompleted! Generated {count} logs to {log_file}")
    
    elif mode == "burst":
        # Generate bursts of logs with pauses
        bursts = int(sys.argv[2]) if len(sys.argv) > 2 else 5
        logs_per_burst = int(sys.argv[3]) if len(sys.argv) > 3 else 50
        
        print(f"Generating {bursts} bursts of {logs_per_burst} logs each")
        
        for burst in range(bursts):
            print(f"\nBurst {burst + 1}/{bursts}...")
            for i in range(logs_per_burst):
                log_entry = generate_log_entry()
                write_log(log_entry, log_file)
                time.sleep(0.01)  # Small delay within burst
            
            if burst < bursts - 1:
                print(f"Waiting 5 seconds before next burst...")
                time.sleep(5)
        
        print(f"\nCompleted! Generated {bursts * logs_per_burst} logs in {bursts} bursts")
    
    else:
        print("Usage:")
        print("  python generate-logs.py batch [count]        - Generate batch of logs (default: 100)")
        print("  python generate-logs.py continuous           - Generate logs continuously")
        print("  python generate-logs.py burst [bursts] [per] - Generate bursts of logs")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/generate-logs.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//DEPS info.picocli:picocli:4.6.3
//DEPS com.google.code.gson:gson:2.10.1
//DEPS org.tinylog:tinylog-api:2.7.0
//DEPS org.tinylog:tinylog-impl:2.7.0
//JAVA 17

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import org.tinylog.Logger;
import picocli.CommandLine;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import picocli.CommandLine.Mixin;
import picocli.CommandLine.Parameters;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.Callable;
import java.util.stream.IntStream;

@Command(name = "generate-logs", mixinStandardHelpOptions = true, version = "1.0",
        description = "Spring Boot JSON Lines Log Generator for Loki + Grafana stack",
        subcommands = {
                GenerateLogs.BatchCommand.class,
                GenerateLogs.ContinuousCommand.class,
                GenerateLogs.BurstCommand.class
        })
class GenerateLogs implements Callable<Integer> {

    static {
        // Configure tinylog format to display on a single line
        System.setProperty("tinylog.writer", "console");
        System.setProperty("tinylog.writer.format", "{date: yyyy-MM-dd HH:mm:ss.SSS} {level} [{thread}] {class} - {message}");
        System.setProperty("tinylog.writer.level", "info");
    }

    private static final Random random = new Random();
    private static final Gson gson = new GsonBuilder().disableHtmlEscaping().create();
    // Removed .env loading; path is provided via CLI mixin.

    public static void main(String... args) {
        System.exit(new CommandLine(new GenerateLogs()).execute(args));
    }

    @Override
    public Integer call() {
        new CommandLine(this).usage(System.out);
        return 0;
    }

    // Picocli mixin to provide a logs path for all commands
    static class LogPathMixin {
        @Option(names = {"-p", "--logs-path"}, paramLabel = "PATH",
                description = "Full path to the log folder (default: ../logs)",
                defaultValue = "../logs")
        String logsPath;
    }

    @Command(name = "batch", mixinStandardHelpOptions = true,
             description = "Generate a batch of log entries")
    static class BatchCommand implements Callable<Integer> {
        @Mixin
        LogPathMixin logPath;
        @Parameters(index = "0", description = "Number of logs to generate", defaultValue = "100")
        private int count;

        @Override
        public Integer call() throws IOException {
            var writer = new LogWriter(logPath.logsPath);
            Logger.info("Generating {} log entries to {}", count, writer.logFile());

            IntStream.range(0, count).forEach(i -> {
                writer.write(LogEntry.generate());
                if ((i + 1) % 10 == 0) {
                    Logger.info("Generated {}/{} logs", i + 1, count);
                }
            });

            Logger.info("Completed! Generated {} logs to {}", count, writer.logFile());
            return 0;
        }
    }

    @Command(name = "continuous", mixinStandardHelpOptions = true,
             description = "Generate logs continuously")
    static class ContinuousCommand implements Callable<Integer> {
        @Mixin
        LogPathMixin logPath;
        @Override
        public Integer call() {
            var writer = new LogWriter(logPath.logsPath);
            Logger.info("Generating logs continuously to {}", writer.logFile());
            Logger.info("Press Ctrl+C to stop");

            try {
                while (true) {
                    var entry = LogEntry.generate();
                    writer.write(entry);
                    ConsoleLogger.log(entry);
                    Thread.sleep((long) (random.nextDouble() * 1900 + 100));
                }
            } catch (InterruptedException e) {
                Logger.info("Stopped log generation");
                return 0;
            }
        }
    }

    @Command(name = "burst", mixinStandardHelpOptions = true,
             description = "Generate bursts of logs with pauses")
    static class BurstCommand implements Callable<Integer> {
        @Mixin
        LogPathMixin logPath;
        @Parameters(index = "0", description = "Number of bursts", defaultValue = "5")
        private int bursts;

        @Parameters(index = "1", description = "Logs per burst", defaultValue = "50")
        private int logsPerBurst;

        @Override
        public Integer call() throws Exception {
            var writer = new LogWriter(logPath.logsPath);
            Logger.info("Generating {} bursts of {} logs each", bursts, logsPerBurst);

            for (int burst = 0; burst < bursts; burst++) {
                Logger.info("Burst {}/{}", burst + 1, bursts);

                for (int i = 0; i < logsPerBurst; i++) {
                    writer.write(LogEntry.generate());
                    Thread.sleep(10);
                }

                if (burst < bursts - 1) {
                    Logger.info("Waiting 5 seconds before next burst...");
                    Thread.sleep(5000);
                }
            }

            Logger.info("Completed! Generated {} logs in {} bursts", bursts * logsPerBurst, bursts);
            return 0;
        }
    }

    record LogEntry(
            String ts,
            String level,
            String type,
            String application,
            String thread,
            String logger,
            String message,
            String traceId,
            String spanId,
            ExceptionInfo exception
    ) {
        static LogEntry generate() {
            var level = LogLevel.random();
            var timestamp = timestamp();
            var thread = ThreadName.random();
            var logger = LoggerName.random();
            var message = Message.generate(level);
            var traceId = random.nextDouble() < 0.5 ? TraceId.generate() : null;
            var spanId = traceId != null ? SpanId.generate() : null;
            var exception = level == LogLevel.ERROR && random.nextDouble() < 0.7
                    ? ExceptionInfo.generate() : null;

            return new LogEntry(timestamp, level.name(), level.name(), "demo-app",
                    thread, logger, message, traceId, spanId, exception);
        }

        private static String timestamp() {
            var bogotaZone = ZoneId.of("America/Bogota");
            var now = ZonedDateTime.now(bogotaZone);
            return now.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSZZZ"));
        }
    }

    record ExceptionInfo(String className, String message, String stackTrace) {
        static ExceptionInfo generate() {
            var exClass = ExceptionClass.random();
            var exMessage = ExceptionMessage.generate();
            var stackTrace = String.format("%s: %s\n\tat com.example.demo.Example.method(Example.java:%d)\n\tat com.example.demo.Main.run(Main.java:%d)",
                    exClass, exMessage, random.nextInt(190) + 10, random.nextInt(90) + 10);

            return new ExceptionInfo(exClass, exMessage, stackTrace);
        }
    }

    enum LogLevel {
        INFO(60), DEBUG(20), WARN(15), ERROR(4), TRACE(1);

        private final int weight;

        LogLevel(int weight) {
            this.weight = weight;
        }

        static LogLevel random() {
            int total = 0;
            for (var level : values()) total += level.weight;

            int value = random.nextInt(total);
            int cumulative = 0;

            for (var level : values()) {
                cumulative += level.weight;
                if (value < cumulative) return level;
            }
            return INFO;
        }
    }

    static class Message {
        private static final Map<LogLevel, List<String>> templates = Map.of(
                LogLevel.INFO, List.of(
                        "Application started successfully",
                        "User logged in successfully",
                        "Order created with ID: {}",
                        "Payment processed successfully",
                        "Database connection established",
                        "Request completed in {}ms",
                        "New user registered: {}",
                        "Session created for user: {}",
                        "Cache refreshed successfully",
                        "Health check passed"
                ),
                LogLevel.DEBUG, List.of(
                        "Entering method: {}",
                        "Exiting method: {}",
                        "Query executed: SELECT * FROM users WHERE id = {}",
                        "Cache hit for key: {}",
                        "Validating request parameters",
                        "Processing request from IP: {}",
                        "Applying security filter",
                        "Deserializing JSON payload",
                        "Loading configuration from: {}",
                        "Initializing bean: {}"
                ),
                LogLevel.WARN, List.of(
                        "Slow query detected: took {}ms",
                        "Cache miss for key: {}",
                        "Deprecated API used: {}",
                        "Retry attempt {} for operation",
                        "Queue size approaching limit: {}",
                        "Connection pool utilization high: {}%",
                        "Session timeout for user: {}",
                        "Invalid input received: {}",
                        "Rate limit approaching for IP: {}",
                        "Configuration value missing, using default"
                ),
                LogLevel.ERROR, List.of(
                        "Failed to process payment: {}",
                        "Database connection error: {}",
                        "Authentication failed for user: {}",
                        "Unable to send email notification",
                        "External API call failed: {}",
                        "Invalid JSON payload received",
                        "Resource not found: {}",
                        "Permission denied for user: {}",
                        "Transaction rollback: {}",
                        "Unexpected exception: {}"
                ),
                LogLevel.TRACE, List.of(
                        "Method trace: {} with parameters: {}",
                        "SQL statement: {}",
                        "HTTP request headers: {}",
                        "Request body: {}",
                        "Response body: {}"
                )
        );

        static String generate(LogLevel level) {
            var template = randomFrom(templates.get(level));
            return formatTemplate(template);
        }

        private static String formatTemplate(String template) {
            if (!template.contains("{}")) return template;

            var result = template;
            while (result.contains("{}")) {
                result = result.replaceFirst("\\{}", generateValue(template));
            }
            return result;
        }

        private static String generateValue(String template) {
            var lower = template.toLowerCase();
            if (lower.contains("id")) return String.valueOf(random.nextInt(9000) + 1000);
            if (lower.contains("ms")) return String.valueOf(random.nextInt(1950) + 50);
            if (lower.contains("user")) return "user" + random.nextInt(100);
            if (template.contains("%")) return String.valueOf(random.nextInt(26) + 70);
            if (template.contains("IP")) return "192.168.1." + (random.nextInt(255) + 1);
            return randomFrom(List.of("alpha", "beta", "gamma", "delta"));
        }
    }

    static class LoggerName {
        private static final List<String> loggers = List.of(
                "com.example.demo.controller.UserController",
                "com.example.demo.service.UserService",
                "com.example.demo.repository.UserRepository",
                "com.example.demo.controller.OrderController",
                "com.example.demo.service.OrderService",
                "com.example.demo.service.PaymentService",
                "com.example.demo.security.AuthenticationFilter",
                "com.example.demo.config.DataSourceConfig",
                "org.springframework.web.servlet.DispatcherServlet",
                "org.springframework.boot.web.embedded.tomcat.TomcatWebServer"
        );

        static String random() {
            return randomFrom(loggers);
        }
    }

    static class ThreadName {
        private static final List<String> threads = List.of(
                "http-nio-8080-exec-1",
                "http-nio-8080-exec-2",
                "http-nio-8080-exec-3",
                "http-nio-8080-exec-4",
                "scheduling-1",
                "task-executor-1",
                "task-executor-2"
        );

        static String random() {
            return randomFrom(threads);
        }
    }

    static class TraceId {
        static String generate() {
            return randomHex(32);
        }
    }

    static class SpanId {
        static String generate() {
            return randomHex(16);
        }
    }

    static class ExceptionClass {
        private static final List<String> classes = List.of(
                "java.sql.SQLException",
                "org.springframework.web.client.HttpClientErrorException",
                "java.io.IOException",
                "java.lang.NullPointerException",
                "javax.validation.ValidationException",
                "org.springframework.security.access.AccessDeniedException",
                "com.example.demo.exception.ResourceNotFoundException",
                "com.example.demo.exception.PaymentException"
        );

        static String random() {
            return randomFrom(classes);
        }
    }

    static class ExceptionMessage {
        private static final List<String> messages = List.of(
                "Connection timeout after 30000ms",
                "Invalid credentials provided",
                "Resource with ID {} not found",
                "Payment gateway returned error code: {}",
                "Validation failed for field: {}",
                "Database constraint violation",
                "API rate limit exceeded",
                "Session expired"
        );

        static String generate() {
            var template = randomFrom(messages);
            return template.contains("{}")
                    ? template.replace("{}", String.valueOf(random.nextInt(9000) + 1000))
                    : template;
        }
    }

    static class LogWriter {
        private final Path logFile;

        LogWriter(String logDir) {
            var logPath = Paths.get(logDir);

            try {
                if (!Files.exists(logPath)) {
                    Files.createDirectories(logPath);
                    Logger.info("Created directory: {}", logDir);
                }
            } catch (IOException e) {
                throw new RuntimeException("Failed to create log directory: " + logDir, e);
            }

            this.logFile = logPath.resolve("application.log");
        }

        void write(LogEntry entry) {
            try {
                var json = gson.toJson(entry) + "\n";
                Files.writeString(logFile, json,
                        StandardOpenOption.CREATE,
                        StandardOpenOption.APPEND);
            } catch (IOException e) {
                Logger.error(e, "Failed to write log");
            }
        }

        String logFile() {
            return logFile.toString();
        }
    }

    static class ConsoleLogger {
        private static final Map<String, String> colors = Map.of(
                "INFO", "\033[32m",
                "DEBUG", "\033[36m",
                "WARN", "\033[33m",
                "ERROR", "\033[31m",
                "TRACE", "\033[37m"
        );
        private static final String reset = "\033[0m";

        static void log(LogEntry entry) {
            var color = colors.getOrDefault(entry.level(), "");
            var formatted = String.format("%s[%-5s] %-50s - %s%s",
                    color, entry.level(), entry.logger(), entry.message(), reset);
            Logger.info(formatted);
        }
    }

    private static <T> T randomFrom(List<T> list) {
        return list.get(random.nextInt(list.size()));
    }

    private static String randomHex(int length) {
        return IntStream.range(0, length)
                .mapToObj(i -> Integer.toHexString(random.nextInt(16)))
                .reduce("", String::concat);
    }
}
